GeoNode/geonode
# [master][Fixes #8112] Dump and restore Geofence rules in case of requests or …

References: #8112

## What was done

* `GeofenceAdapter` and  `GeofenceLayerRulesUnitOfWork` were created:
    * When inside the `GeofenceLayerRulesUnitOfWork` the geofence transactions were stacked to be executed at the end (`__exit__`).
*  `GeofenceLayerRulesUnitOfWork` were used as a context manager.
* In case of failure:
    * `geofence_adapter.rollback()` restore the previous state, if needed. 

## How to test

* Django Testing
```
./manage.py test geonode.security.tests.PermissionsTest --keepdb
./manage.py test geonode.security.tests.BulkPermissionsTests --keepdb
```

* Manual Testing:
    * Run the application.
    * Add a layer.
    * Change its permissions:
    ![image](https://user-images.githubusercontent.com/12943132/133075679-98127ad1-34fa-49e8-b67c-39a6b18ba7d1.png)
    * During execution, check the layer rules count using the following geoserver request:
    ```bash
    # Change `localhost:8080` to your geoserver url, `Basic YWRtaW46Z2Vvc2VydmVy' to your geoserver authorization token and `san_andres_y_providencia_coastline0` to your layer name.
    curl --location --request GET 'http://localhost:8080/geoserver/rest/geofence/rules?workspace=geonode&layer=san_andres_y_providencia_coastline0' \
    --header 'Authorization: Basic YWRtaW46Z2Vvc2VydmVy'
    ```
    * It should reflect the changed you made using the geonode interface.
* For manual testing the rollback:
  * Try to raise an Exception after the Unit of Work block (`geonode/security/models.py#L422`, `PermissionLevelMixin.set_permissions`).
      ```python
                                  self.set_dirty_state()
                  raise Exception()
              except Exception as e:
                  geofence_adapter.rollback()
                  raise GeoNodeException(e)
      ``` 
  * Verify at the geoserver (same cURL as above) if the rules list has been modified during the execution and restored at the end.
  * At the geonode interface, you will have the following message:
  ![image](https://user-images.githubusercontent.com/12943132/133076986-c42798d8-64c5-4c20-a276-c0a570bcee72.png)


## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:


The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**
